### ISSUE TITLE
SecureRequest: Can't connect to sessionserver.mojang.com: SSL handshake failed: -0x2700

### ISSUE THREAD

**Issue by Michael-01** [cuberite/cuberite#2115]
Since this week we are facing that issue.

SecureRequest: Can't connect to sessionserver.mojang.com: SSL handshake failed: -0x2700
Kicking player thessy09 for "Failed to authenticate account!"

Before our MCServer running on a Raspberry worked fine.
I installed that MCServer for my kids, so that are able to play in one world together in the Home-LAN 
I did not change anything. 

Does anyone now how to fix that issue.
Thanks a lot in advance for your support 
br. Michael

**Comment by bearbin** [comment]
You need to upgrade to the latest version by downloading the latest build. This was fixed by #2085.

**Comment by Michael-01** [comment]
Thanks for your reply .

What exactly do I need to do.
Could you please describe that in more detail. I am not so experienced 
Thank you very much.

**Comment by bearbin** [comment]
What method did you use to first install MCServer?

**Comment by Michael-01** [comment]
Sorry
A body of mine did install the MCServer on the Pi, maybe I need to ask him to install the latest build.
I guess you are talking about to re-install the MCserver new with the latest build.

I even don't know what version installed.  I guess it is build id: #80 what ever thats mean

[15:37:20] --- Started Log ---

     [15:37:20] MCServer Jenkins gcc armhf Release build id: #80
     [15:37:20] from commit id: 31953b19b805facde2114855563540a81bf9bdd8 built at: Sun Apr  5 11:41:45 EDT 2015
     [15:37:20] Creating new server instance...
     [15:37:20] Reading server config...
     [15:37:20] Starting server...
info [15:37:20] Compatible clients: 1.7.x, 1.8
info [15:37:20] Compatible protocol versions 4, 5, 47

**Comment by bearbin** [comment]
Yes. If you asked your friend to install the latest build for you that would be good.

**Comment by Michael-01** [comment]
Hi bearbin,

Thanks for your help so far.
Do you know the most actual version of the MC Jenkins?
Thanks again

**Comment by bearbin** [comment]
The latest version is #25.

Download is here: http://builds.cuberite.org/job/MCServer%20Linux%20armhf/lastSuccessfulBuild/artifact/MCServer/MCServer.tar.gz

**Comment by Michael-01** [comment]
Thank you!

**Comment by JackCarterSmith** [comment]
I found a new bug with invalid SSL handshake on both Linux and Window release (351).

![firefox_2022-11-06_12-02-55](https://user-images.githubusercontent.com/2466196/200167047-53ab5a38-cd11-479e-ade8-ee8804e88741.png)

Need to update certs again?

**Comment by tigerw** [comment]
Yep, #5456.

**Comment by JackCarterSmith** [comment]
Nice! 💯

**Comment by tonitch** [comment]
It seems like the problem is clearly not solved yet!
there is a lot of people complaining about the authentification that doesn't work.
It doesn't work for me and I can't see no issue open about this currently... maybe we should reopen this one (or redo one I don't know)

for my part, I don't think I have the coding skill and knowledge to do anything by myself but I'm willing to help by providing any info you would need or testing anything you would need to test ^^

hope this could be fixed tho! I think that if this last for too long this will definitely kill the everlasting last breath of cuberites project :cry: 

:crossed_fingers:

**Comment by JackCarterSmith** [comment]
I've tested it last month and it's working now from my side. But yeah, I share the same feeling about the life of Cuberite...
For that issue of certificats, I'm not sure if a dynamic auto-updater can do the job. I didn't have the strength (and skills maybe? I'm just an hardware engineer after all) to take a look more deeply in the code...

@tonitch You have always auth problems with mojang certs?

**Comment by tonitch** [comment]
Well, I have auth problem on cuberite since quite some time now yhea, but I'm pretty sure it's the same time as your last message here ^^

I made a quick look trying to find what could be wrong in the code and I found in https://github.com/cuberite/cuberite/blob/8a763d3bedac3eabee9c1ca022be53038ba3fc54/src/Protocol/Authenticator.cpp#L21-L22

I tried to curl this link with my name and I can't get response.... so I look it up on wiki.vg and I can't seem to find this link anywhere... maybe the auth sequence has just changed and that would explain why 

but I've never looked into mojang's auth sequence so I could be wrong tho! hope this can help ?

**Comment by Janmm14** [comment]
@tonitch here the wiki.vg page containing the urls: https://wiki.vg/Protocol_Encryption

**Comment by tonitch** [comment]
Ho right, i'm sorry I though I found something there... seems like the search function of wiki.vg is not so reliable ^^

still, I hope we can find something ^^

**Comment by tigerw** [comment]
> there is a lot of people complaining about the authentification that doesn't work

Is this in Discord? What was the error they were getting?

**Comment by 12xx12** [comment]
The -0x4c error

**Comment by tonitch** [comment]
well not everyone has sent the error (I didn't think about asking because I tough it was the same error for everyone... sorry...) Indeed, the last one that asked say that his error was -0x4c.

On my side tho, I still have -0x2700, which lead me to think that this is a related issue to the original post here.
I tried to look a bit what was happening but I don't really know a lot about ssl so I couldn't figure things out... I even tried changing certificates with my own but it didn't worked either. (I could have done it wrong tho)

**Comment by mathiascode** [comment]
> On my side tho, I still have -0x2700, which lead me to think that this is a related issue to the original post here. I tried to look a bit what was happening but I don't really know a lot about ssl so I couldn't figure things out... I even tried changing certificates with my own but it didn't worked either. (I could have done it wrong tho)

Are you sure you're on the latest build? The certificate is correct, so there shouldn't be a handshake error anymore. I only see the  -0x4c error (MBEDTLS_ERR_NET_RECV_FAILED) since 8a763d3bedac3eabee9c1ca022be53038ba3fc54.

**Comment by tonitch** [comment]
well this is my verry bad.... I worked on my endermite branch and so forgot to pull from origin to make my tests...

Indeed the problem is still there but the error code is -0x4c

**Comment by mathiascode** [comment]
Similar findings as https://github.com/cuberite/cuberite/pull/5456#issuecomment-1321197243 after looking into this a bit.

This code is called until all incoming data is processed:

https://github.com/cuberite/cuberite/blob/b0376367562d6c2c9f69d3b16b3c62bf7c353c78/src/mbedTLS%2B%2B/BlockingSslClientSocket.cpp#L278-L299

This code is called, `m_IsConnected` becomes false:

https://github.com/cuberite/cuberite/blob/b0376367562d6c2c9f69d3b16b3c62bf7c353c78/src/mbedTLS%2B%2B/BlockingSslClientSocket.cpp#L366-L371

Then, `cBlockingSslClientSocket::ReceiveEncrypted` is called again. `a_NumBytes` has a value of 5, but `m_IncomingData` is empty:

https://github.com/cuberite/cuberite/blob/b0376367562d6c2c9f69d3b16b3c62bf7c353c78/src/mbedTLS%2B%2B/BlockingSslClientSocket.cpp#L289-L292

@tigerw Any idea why this hasn't been an issue until now?

**Comment by tigerw** [comment]
I think something changed in the communication between us and the session server, which caused the TLS flow to subtly change, and it exposed a bug in our implementation. Could be something like the certificate sizes got bigger/smaller meaning the order of reads changed, etc.

**Comment by NiLSPACE** [comment]
I started looking into this. Weirdly enough all the data gets in the [buffer](https://github.com/cuberite/cuberite/blob/8a763d3bedac3eabee9c1ca022be53038ba3fc54/src/Protocol/MojangAPI.cpp#L480). Only after the buffer is filled the [while loop](https://github.com/cuberite/cuberite/blob/8a763d3bedac3eabee9c1ca022be53038ba3fc54/src/Protocol/MojangAPI.cpp#L456) goes again. Because of that the [cBlockingSslClientSocket::ReceiveEncrypted](https://github.com/cuberite/cuberite/blob/5ae924ec74bab199284c451552040bcc6f39bb3a/src/mbedTLS%2B%2B/BlockingSslClientSocket.cpp#L278) callback is called again.  Because all the data has been received already `m_IncomingData` is empty. I suspect the reason why it doesn't always fail is because sometimes the server doesn't close the connection fast enough to reach the [if statement](https://github.com/cuberite/cuberite/blob/5ae924ec74bab199284c451552040bcc6f39bb3a/src/mbedTLS%2B%2B/BlockingSslClientSocket.cpp#L289) causing both `m_IsConnected` to be `false` and `m_IncomingData` to be empty.

Adding 
```cpp
if ((ret >= 0) && (ret < sizeof(buf)))
{
	break;
}
```
after [appending the buffer](https://github.com/cuberite/cuberite/blob/8a763d3bedac3eabee9c1ca022be53038ba3fc54/src/Protocol/MojangAPI.cpp#L480) does technically fix it, however I don't know mbedtls well enough to know if `mbedtls_ssl_read` can return a partially filled buffer while the connection hasn't completed yet.

**Comment by GILLESMaster** [comment]
I would love to know when this gets fixed, because this error is super annoying, for me I just disabled auth en used a proxy, but it isn't my intended way of hosting.

**Comment by 12xx12** [comment]
We are way behind on the mbedtls. Maybe try updating?

**Comment by 12xx12** [comment]
> ```cpp
> if ((ret >= 0) && (ret < sizeof(buf)))
> {
>	break;
>}
>```

What does the ret value represent in this case? (I lost track in the `mbedtls_ssl_read ` method)

**Comment by NiLSPACE** [comment]
`ret` is the amount of bytes read that is now in the buffer.

If it's negative it means an error has occurred, so my "solution" wouldn't work.

Another option is to remove the `m_IsConnected` and `m_IncomeData` check: https://github.com/cuberite/cuberite/blob/5ae924ec74bab199284c451552040bcc6f39bb3a/src/mbedTLS%2B%2B/BlockingSslClientSocket.cpp#L289-L292

Instead of returning an error code when that occurs it would just continue and [copy nothing in the buffer](https://github.com/cuberite/cuberite/blob/5ae924ec74bab199284c451552040bcc6f39bb3a/src/mbedTLS%2B%2B/BlockingSslClientSocket.cpp#L295-L297). In that case `ret` would be `0` and the [connection would close anyway](https://github.com/cuberite/cuberite/blob/8a763d3bedac3eabee9c1ca022be53038ba3fc54/src/Protocol/MojangAPI.cpp#L475).

**Comment by 12xx12** [comment]
But [here](https://github.com/cuberite/cuberite/blob/fddbf65e2819d487021946a51722d23565bdd561/src/Protocol/MojangAPI.cpp#L470) any negative value is already caught

**Comment by NiLSPACE** [comment]
Yes, but there it returns false which causes `cAuthenticator` to see it as a player that failed to authenticate: https://github.com/cuberite/cuberite/blob/1d72306bae40c38a8a684cc412b0478fb0920d63/src/Protocol/Authenticator.cpp#L163-L166

**Comment by 12xx12** [comment]
Yes, but with your solution that path isnt called

**Comment by 12xx12** [comment]
Oh. Now I see. Hm. Yeah, maybe remove that. Maybe @madmaxoft can tell us more. They wrote that code

**Comment by NiLSPACE** [comment]
Yeah, because I'm not sure [this](https://github.com/cuberite/cuberite/blob/5ae924ec74bab199284c451552040bcc6f39bb3a/src/mbedTLS%2B%2B/BlockingSslClientSocket.cpp#L289-L292) is an error. (Just to be clear, I'm not knowledgeable enough about TLS and especially mbedtls to be sure)

All I can see from debugging is that the whole reply is already in [`a_Response`](https://github.com/cuberite/cuberite/blob/8a763d3bedac3eabee9c1ca022be53038ba3fc54/src/Protocol/MojangAPI.cpp#L480)  and that for some reason the for loop goes again. Because of that the [`cBlockingSslClientSocket::ReceiveEncrypted`](https://github.com/cuberite/cuberite/blob/5ae924ec74bab199284c451552040bcc6f39bb3a/src/mbedTLS%2B%2B/BlockingSslClientSocket.cpp#L278) callback gets called again, this time without any incoming data because everything has already been received which causes [the check](https://github.com/cuberite/cuberite/blob/5ae924ec74bab199284c451552040bcc6f39bb3a/src/mbedTLS%2B%2B/BlockingSslClientSocket.cpp#L289-L292) to fail.

**Comment by 12xx12** [comment]
So we need to leave the loop one step early. Or find the reason the loop goes again

**Comment by 12xx12** [comment]
Why do we even bother and not use the cURLClient class?

**Comment by 12xx12** [comment]
I am using cUrlClient now and it works

**Comment by Bixilon** [comment]
Its happening (again?): `[14:59:53] SecureRequest: SSL reading failed: -0x4c`
